crates/solver/src/settlement_rater.rs

```diff
@@ -259,3 +277,156 @@ impl SettlementRating for SettlementRater {
         Ok(Rating::Ok(rated_settlement))
     }
 }
+
+pub struct ScoreCalculator {
+    score_cap: BigRational,
+}
+
+impl ScoreCalculator {
+    pub fn new(score_cap: BigRational) -> Self {
+        Self { score_cap }
+    }
+
+    pub fn compute_score(
+        &self,
+        objective_value: &BigRational,
+        success_probability: f64,
+    ) -> Result<Score> {
+        ensure!(
+            (0.0..=1.0).contains(&success_probability),
+            "success probability must be between 0 and 1."
+        );
+        let success_probability = BigRational::from_float(success_probability).unwrap();
+        let cost_fail = BigRational::from_float(0.0).unwrap();
+        let optimal_score =
+            self.compute_optimal_bid(objective_value.clone(), success_probability, cost_fail)?;
+        let score = big_rational_to_u256(&optimal_score).context("Invalid score.")?;
+        Ok(Score::Protocol(score))
+    }
+
+    fn compute_optimal_bid(
+        &self,
+        objective: BigRational,
+        probability_success: BigRational,
+        cost_fail: BigRational,
+    ) -> Result<BigRational> {
+        tracing::trace!(
+            ?objective,
+            ?probability_success,
+            ?cost_fail,
+            "Computing optimal bid"
+        );
+        let probability_fail = BigRational::one() - probability_success.clone();
+        // filter out invalid solution where there is no chance of success or where the
+        // success is 100% (because it can't be true, there is always some risk of
+        // failure)
+        ensure!(!probability_success.is_zero(), "success is zero");
+        ensure!(!probability_fail.is_zero(), "fail is zero");
```

Comment:
removed checks but used `checked_div` for safety reasons. other code could change and in the future it might not be guaranteed that these values are not zero.